dvc/scm/git.py

```diff
@@ -491,3 +497,160 @@ def resolve_commit(self, rev):
+        for backend in self.backends:
+            try:
+                func = getattr(backend, name)
+                return func(*args, **kwargs)
+            except (AttributeError, NotImplementedError):
+                pass
+        raise NoGitBackendError(name)
+
+    _ignored = partialmethod(_backend_func, "_ignored")
+    set_ref = partialmethod(_backend_func, "set_ref")
+    get_ref = partialmethod(_backend_func, "get_ref")
+    remove_ref = partialmethod(_backend_func, "remove_ref")
+    push_refspec = partialmethod(_backend_func, "push_refspec")
+    fetch_refspecs = partialmethod(_backend_func, "fetch_refspecs")
+    _stash_iter = partialmethod(_backend_func, "_stash_iter")
+    _stash_push = partialmethod(_backend_func, "_stash_push")
+    _stash_apply = partialmethod(_backend_func, "_stash_apply")
+    reflog_delete = partialmethod(_backend_func, "reflog_delete")
+
+    def get_refs_containing(self, rev: str, pattern: Optional[str] = None):
+        """Iterate over all git refs containing the specfied revision."""
+        from git.exc import GitCommandError
+
+        try:
+            if pattern:
+                args = [pattern]
+            else:
+                args = []
+            for line in self.repo.git.for_each_ref(
+                *args, contains=rev, format=r"%(refname)"
+            ).splitlines():
+                line = line.strip()
+                if line:
+                    yield line
+        except GitCommandError:
+            pass
+
+    @contextmanager
+    def detach_head(self, rev: Optional[str] = None):
+        """Context manager for performing detached HEAD SCM operations.
+
+        Detaches and restores HEAD similar to interactive git rebase.
+        Restore is equivalent to 'reset --soft', meaning the caller is
+        is responsible for preserving & restoring working tree state
+        (i.e. via stash) when applicable.
+
+        Yields revision of detached head.
+        """
+        if not rev:
+            rev = "HEAD"
+        orig_head = self.get_ref("HEAD", follow=False)
+        logger.debug("Detaching HEAD at '%s'", rev)
+        self.checkout(rev, detach=True)
+        try:
+            yield self.get_ref("HEAD")
+        finally:
+            prefix = os.fsdecode(self.LOCAL_BRANCH_PREFIX)
+            if orig_head.startswith(prefix):
+                orig_head = orig_head[len(prefix) :]
+            logger.debug("Restore HEAD to '%s'", orig_head)
+            self.checkout(orig_head)
+
+    @contextmanager
+    def stash_workspace(self, **kwargs):
+        """Stash restore any workspace changes.
+
+        Yields revision of the stash commit. Yields None if there were no
+        changes to stash.
+        """
+        logger.debug("Stashing workspace")
+        rev = self.stash.push(**kwargs)
+        try:
+            yield rev
+        finally:
+            if rev:
+                logger.debug("Restoring stashed workspace")
+                self.stash.pop()
+
+
+class Stash:
```

Comment:
Feels like we could make git.py -> git/__init__.py and have backend in git/backend/* and stash in a git/stash.py.

Reply:
Btw, stash is purely git concept, right? Meaning that we don't ever need it for no-scm case and other things like mercurial don't have it? Talking about hg just theoretically, there is no plans to support hg right now, but the arch here assumes that there might be alternative scm implementations.

Reply:
It's been a while since I used hg for anything, but yeah I think there's no stash equivalent out of the box for hg.